### DISCUSSION TITLE
Testing release 4.4. Testers needed.

### DISCUSSION THREAD

**Discussion by jantje** [Sloeber/arduino-eclipse-plugin#1290] [Q&A]
Hi,
As you may have noticed we disabled the nightly because there is some disruptive changes in Sloeber.
We would like some testing to happen before we release 4.4.
The testing focusses on user experience and migration of old workspaces.
Testing is planned to start feb 2021 and we are looking for people willing to test.

please respond below (and subscribe to this discussion to be updated) if you want to help out with the testing.

Thanks Jantje

Why this extra test is needed.
As you may know we developed quite extensive testing for Sloeber so why is there extra manual testing needed you may ask.
Well "data migration" is hard to test because data is not easily readily available. Your workspaces is the data we are looking for.
Secondly because we moved the data around (more info bellow) that may have side effects that are not covered in our testing. We are mostly thinking of GUI stuff no longer working (hard to test as our tests run without GUI).
Also some new "use cases" have been added for which little or no testing is available.

So what are these changes?
1) Sloeber used to store all its data in CDT managed environment variables. Now Sloeber stores its data in a .sproject file and has its own environment variable provider. This has huge implications in regards to configuration management. So if you have multiple configuration project; please sign up for testing.

2) Sloeber used to convert environment varriables to uppercase. This was so because CDT did that. We fixed CDT so version 4.4 needs CDT10 or higher where previous versions of Sloeber have not been tested with CDT10 and may not work (probably will as it is workiong in linux which did not uppercase in CDT)

3) Java version upgraded. Sloeber product on windows came with a jrte folder containing jave 8. Sloeber 4.4 (and the current nightly) does not work with java8. Java 11 is required (I think java15 causes CDT issues).

4)Better support for version control. For each configuration you can specify whether or not you want it in version control (project properties->arduino->other) If one config is marked as "in version control" a sloeber.cfg file is created. When this file changes (for instance because of a get or your editing) Sloeber auto magically applies the changes. This means that from now on you no longer need to check in the .setting and other . files. Also the sloeber.cfg files are suposed to be platform independent. As sutch they are in / format and do not contain platform dependent info like FQN for boards location (if the boards are managed by sloeber boards manager) and no upload ports are in this file.

5) a convert to sloeber project wizard. When checking out a project from version control you should check out as a "standard project" because there is no .project and no .cproject and no .sproject. Then you can go to new wizards->arduino and convert to sloeber project.

6) Sloeber workarounds moved mostly from internally to externally. This means that Sloeber maintains boards.sloeber.txt and platform.sloeber.txt files with most/all the workarounds in there.

**Test cases**

0) Copy/Zip your current workspace(s) so that the steps can be repeated (and your current work is saved).
1) Convert a Sloeber 3.3 install to a Sloeber 3.4 install. (It might be it is simply not working but we want to know "how well this will work")
2) Convert a Sloeber 3.3 project into a Sloeber 3.4 project
3) Convert a Sloeber 3.3 workspace in a Sloeber 3.4 workspace
4) check-in a sloeber project in version control and check it out on another system (same os and different os)
5) test project creation
5.1) single configuration
5.2)multiple configuration
6) test project build
6.1) single configuration
6.2)multiple configuration
7) test project upload
7.1) single configuration
7.2)multiple configuration

**Reporting issues:**
Results should be reported as issue. We propose to use following notation for issue description
Release testing 4.4 test case [test case number] os:[win/mac/linux]
In case of a successful test and no successful entry exists
-> add a entry with description success 
If an entry like that already exists
-> add thumbs up to the success entry

In case of a failed test and no entry exists describing your problem
-> create a new entry. If the same problem exists on other os's link to that entry
In case of a failed test and no entry exists describing your problem
-> add a thumbs down to the entry describing your problem

**Comment by erdemuncuoglu** [comment]
I can test Sloeber on linux (amd64) with boards ESP32, ESP8266 and Arduino UNO.

**Comment by uzi18** [comment]
Also can test linux 64bit build, maybe also win64.

**Comment by robamu** [comment]
I can test it, but just some basic stuff with Uno, Nano Every and Due, I have not done any complex projects with Sloeber yet. I can test on Linux (Ubuntu 20.04) and Windows 10, amd64

**Reply by jantje** [reply]
Welcome :-) The more the better :-)

**Comment by hreintke** [comment]
@jantje 
I can test on Windows 10, using esp8266, esp32, stm32.
Using boardmanager and private hardware paths.

Questions : 
From where do you want me to download 4.4 ?
Are there specific testcases that you want to have tested ?
Where/how would you like to receive the feedback from testing ?

**Reply by jantje** [reply]
I added some test cases and a reporting strategy to the first entry. Please feel free to add/refuse/counter propose...
A release will be made this month (feb 2021). I'm checking the code for TODO's and TOFIX things and I need to run the tests on linux. (I do not own a mac)
I will make a beta release in this github repository and if needed a update site on baeyens.it.

**Comment by rlogiacco** [comment]
Count me in sir: Windows 10 x64, Italian localization, mostly Arduino boards (pro micro via usbasp, micro with board reset) and ESP32

**Comment by uzi18** [comment]
@jantje maybe we can add tag to issues for beta_4.4

**Reply by jantje** [reply]
Sounds like a good idea

**Comment by jantje** [comment]
I released the first version for testing.
https://github.com/Sloeber/arduino-eclipse-plugin/releases/tag/V4_4_0-beta
My first look on upgrading was not looking good :-(

**Comment by jantje** [comment]
I fixed all issues untill now.
Shall I make a new release or is someone having some other issue somewhere?

**Comment by KlausJoergensen** [comment]
Hi Jantje,
I tranfer my data between Ubuntu and Win10. Shall I update both systems to nightly? Ubuntu has no problems until now.

**Reply by jantje** [reply]
There is no nightly for 4.4. Only the project install

**Comment by robzed** [comment]
Tried on sloeber-ide-sloeber_release-macosx.cocoa.x86_64.zip from https://github.com/Sloeber/arduino-eclipse-plugin/releases/tag/V4_4_0-beta

OS: MacOS Big Sur Version 11.2.1, Machine: MacBook Pro Mid-2015.

Had to make these changes:
sudo chmod -R 755 Sloeber.app 
sudo xattr -r -d com.apple.quarantine Sloeber.app
Into Contents/Eclipse/sloeber-ide.ini
  -vm
  /Users/rob/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.macosx.x86_64_15.0.1.v20201027-0507/jre/lib/libjli.dylib

Fresh install, no other Sloeber. (I have an Eclipse install 2020-12).

Project not before with Sloeber had Sloeber added, and it builds. 
Upload:
  Com port not found. Skipping reset 
  Cannot run program <...> avrdude
(I will look at that, I suspect it's configuration).

**Reply by jantje** [reply]
>sudo chmod -R 755 Sloeber.app
Did you unpack using the macOS native Archive Utility?

>sudo xattr -r -d com.apple.quarantine Sloeber.app

As described here http://eclipse.baeyens.it/stable.php?OS=MacOS

>Into Contents/Eclipse/sloeber-ide.ini
-vm
/Users/rob/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.macosx.x86_64_15.0.1.v20201027-0507/jre/lib/libjli.dylib

I guess that is to get java11 or higher which is not the default on your system.

>Cannot run program <...> avrdude

This may be the same as https://github.com/Sloeber/arduino-eclipse-plugin/issues/1301
Guess I'd better make a new release

**Reply by robzed** [reply]
> Did you unpack using the macOS native Archive Utility?

Yes

>As described here http://eclipse.baeyens.it/stable.php?OS=MacOS

Yes :-) Been there :-)

>I guess that is to get java11 or higher which is not the default on your system.

That line was from my 2020-12 Eclipse.ini file. I think I installed it with the new Eclipse installer - although I had a problem with one machine with this installer (I think it was a different windows machine). I didn't tweak the Eclipse.ini file at all... unlike in the good old days of Eclipse.

I have Oracle Java installed and it says it's up to date, but I really don't know what Oracle and Apple is doing. And although I've written bits of Java, I'm not really a Java guy and and haven't shipped Java products into the wild - so I'm guessing a lot of the time. 

From terminal this is what I get:
% java -version
java version "1.8.0_281"
Java(TM) SE Runtime Environment (build 1.8.0_281-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.281-b09, mixed mode)

If you need me to check anything else, let me know.

> "This may be the same as #1301"

Thanks I'll take a look.


> Guess I'd better make a new release

I'll give it a whirl if that helps.

**Reply by robzed** [reply]
I wonder if the Eclipse installer hacks past the outdated system default shown in the terminal.

**Reply by jantje** [reply]
Your systems defaults to java 8 -which is old and no good for Sloeber-. If I understood correctly there are some java problems which makes java is no longer updated/installed on mac.
The eclipse installer installs java and the line you added to the eclipse.ini is to tell it to use java installed at xxx. I assume that it is the install location of the openjdk java installed by eclipse installer.

**Comment by jantje** [comment]
A  new beta release is available at https://github.com/Sloeber/arduino-eclipse-plugin/releases/tag/V4_4_0-beta2

**Reply by robzed** [reply]
Tried on my Mac. Works (after chmod/xattr/vm line). AVRDude still fails. Is the change for  #1301 in this build?

    Launching: "/Users/rob/eclipse/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude" "-C/Users/rob/eclipse/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf" -patmega328p -carduino "-P/dev/cu.usbserial-14240" -b115200 -D "-Uflash:w:/Users/rob/Current_Projects/ukmarsey/Release/ukmarsey.hex:i" 
    Output:
    Cannot run program ""/Users/rob/eclipse/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude"" (in directory "/Users/rob/Current_Projects/sloeber_ukmarsey_workspace"): error=2, No such file or directory
    failed to upload "L/ukmarsey/${workspace_loc:/ukmarsey}/Release/ukmarsey.hex"

The line from "Launching" runs from the command line (once I set the Port).

**Reply by jantje** [reply]
It does look like #1301 but that should be fixed.
Are you sure you have the latest version?
![afbeelding](https://user-images.githubusercontent.com/1175841/108636208-144c9380-7484-11eb-9512-a417e95e507e.png)

PS I see the version is wrong :-( but the date is always right

**Reply by robzed** [reply]
![image](https://user-images.githubusercontent.com/4871013/108637993-02b7bb80-7485-11eb-8636-6e1208d80143.png)

**Reply by jantje** [reply]
That looks fine... why does it fail?
Need to think

**Reply by robzed** [reply]
I don't think it's double double quotes. This works from command line:

""/Users/rob/eclipse/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude"" "-C/Users/rob/eclipse/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/etc/avrdude.conf" -patmega328p -carduino "-P/dev/cu.usbserial-14240" -b115200 -D "-Uflash:w:/Users/rob/Current_Projects/ukmarsey/Release/ukmarsey.hex:i"

**Reply by robzed** [reply]
There are two items at least two items I can think it might be:
1. The shell is stripping off the double quotes,and the input to ProcessBuilder has doubl quotes that are not being stripped off by anything, therefore file not found
2. Something in MacOS is not allowing the external tool to run from Eclipse (security measure of some sort).

To test this, I did the following - and I believe it is the quotes (basically me revoking previous post!) and it's likely item 1 not item 2 (since I guess External Tools are run within the context of Eclipse. 
![with_quotes](https://user-images.githubusercontent.com/4871013/108772391-da978d80-7554-11eb-9eca-01351bc43570.png)
![no_quotes](https://user-images.githubusercontent.com/4871013/108772400-dd927e00-7554-11eb-8996-abc7d67a97e3.png)

**Reply by jantje** [reply]
ok I got it.
It seems #1301 was not properly reverted. My bad, Didn't check properly.

**Comment by jantje** [comment]
I did not create a new release but I updated the downloads in https://github.com/Sloeber/arduino-eclipse-plugin/releases/tag/V4_4_0-beta2

**Reply by robzed** [reply]
Thank you. That works!

    avrdude: verifying ...
    avrdude: 12174 bytes of flash verified

    avrdude done.  Thank you.

    The execution of command "avrdude" is done.

I will carry on using this for all my Arduino development ... and give any interesting feedback I find.

**Comment by jantje** [comment]
How is testing going?
Do you think it is ok to release to the bigger public?

**Reply by robzed** [reply]
So I've been using it for about  19 days - on and off in my spare time. I've done quite a bit of robot development on it. I've 100% switched to Sloeber for the low level CPU (Arduino Nano, about 16K of code including PIDs, CLI, and more). Everything from coding, compiling and programming the Nano over USB is via Sloeber. (The upper level CPU is not Arduino but PiZero - and at the moment I'm running a separate Eclipse 2020-12 instance with PyDev installed). 

Sloeber is on my machine constantly - I have experienced no problems. The uptime on this MacBook Pro is 8 days - because of a Mac OS update. (I'm running Big Sur 11.2.2 - there 11.2.3 update waiting). 

I think it's suitable for release - but I haven't tested a lot of Arduinos or a lot of machines. (I do have a Arduino 33 BLE somewhere ...).


Regards,
Rob

**Comment by jantje** [comment]
I made a new issue. Seems all newly created projects may have problems when using libraries
https://github.com/Sloeber/arduino-eclipse-plugin/issues/1316

**Comment by SturmWolfgang** [comment]
I can make a test runs too.

But I do not fully understand what is meant by the different test scenarios.
I do not have a programmer background and do this in spare time only.

> _1. convert a Sloeber 3.3 install to a Sloeber 3.4 install. (It might be it is simply not working but we want to know "how well this will work")_

 
how to do this? if I follow the procedure of building Sloeber 3.4 out of the gitHub repository I have a independend version. 
How to proceed for an upgrade?

> _2. Convert a Sloeber 3.3 project into a Sloeber 3.4 project_

how to? just copy the project to the workspace of 3.4 and open it or are there special commands for that?

> _3. Convert a Sloeber 3.3 workspace in a Sloeber 3.4 workspace_

copy it and open or special procedure?

> _5. test project creation
>    5.1) single configuration
>    5.2)multiple configuration_

whats meant by single/multiple configuration? 

Just for someone willing to test 3.4 here a summary how I got it working on Windows10 64bit:
- install maven: [http://maven.apache.org/download.cgi](url)
- install git: [http://git-scm.com/downloads](url)
- install Java 11: [https://www.computerbase.de/downloads/systemtools/entwicklung/java-se-development-kit/](url)
  (I need to reboot to take effect)
- clear every path information of "java" from the environment settings and but the java 11 bin folder instead, for me:
    `c:\Program Files\Java\jdk-11.0.10\bin\`
    (without it I got errors during built and/or running sloeber, that my current version is not sufficient)
- download git repository in subfolder "sloeber":
    `git clone https://github.com/Sloeber/arduino-eclipse-plugin sloeber`
- build it via maven (for your current os/plattform)
	`mvn clean verify -DskipTests=true`
- if I run win32x64.cmd I got warning regarding too long path
- therefor I copy the build folder to a path shorter than 40 characters 
     its here: \io.sloeber.product\target\products\
- start sloeber by sloeber-ide.exe

**Reply by jantje** [reply]
Thanks for wanting to help out.

_1. convert a Sloeber 3.3 install to a Sloeber 3.4 install. (It might be it is simply not working but we want to know "how well this will work")_

This is simply upgrading a Sloeber 3.3 install to 3.4 install which can now be done because the nightly contains 3.4.
This will automatically trigger test 2. So take backups from your installs and workspaces.

_2. Convert a Sloeber 3.3 project into a Sloeber 3.4 project_

This means that you take a project from a Sloeber 3.3 workspace (so a arduino project you developed in Sloeber 3.3) and try to open the workspace with Sloeber 4.4. This can be done by upgrading your 3.3 install (which is test 1) or by installing 3.4 and opening the workspace in 3.4.

_whats meant by single/multiple configuration?_

When you create a sloeber project it is a single configuration project. The configuration is called Release and that is why after a build you have a folder named Release in your project.
You can add other configurations like debug, Mega, ESP32, WithoutLCD, ....
They all work on the same code (you can exclude code from a config) but can be completely different in compile options (like toolchain/board) and since 3.4 the add to the sloeber.cfg file option.
To create a new config; goto project properties->arduino and select manage configurations. (There are more entries to get to the manage configurations but this one is tested and you probably want to be in the project properties to change settings anyway)
You can select the "active configuration" by right clicking on the project.

**Reply by SturmWolfgang** [reply]
for 1.) I still need smaller steps description: how can I update a 3.3 to 3.4? just copying the files of 3.4 over an existing 3.3 or is there a function to do so?

**Reply by jantje** [reply]
The process is documented here http://eclipse.baeyens.it/how_to.shtml#/n

**Comment by ArminJo** [comment]
Please add a file Upgrade_Instruclions.txt or likewise with this or similar content to the Release.

Upgrade to a new Sloeber version
Copy files from the old to the new Sloeber directory before running it the first time.
1. All files from Sloeber/configuration/settings/prefs
2. jargon-wordlist.txt
3. The complete arduinoPlugin directory
If you have some strange errors for the new builds, sometimes it helps to change one setting in Project > Properties > Arduino.

**Comment by jantje** [comment]
After many changes to make the multiple config reliable and consistent in behaviour I bit the bullet and released 4.4
Thanks for the testing/help
Jantje

PS I made a video on the version control improvements. If there are comments, it is at the time of writing still hidden https://www.youtube.com/watch?v=bqdw1L_8Gjs
